lib/fluent/plugin/in_cloudwatch_ingest.rb

```diff
@@ -279,56 +284,60 @@ def run
+          begin
+            # Fetch log group streams and emit the events
+            log_streams(group, @log_stream_name_prefix).each do |stream|
+              state.store[group][stream] = {} unless state.store[group][stream]
+
+              log.info("processing stream: #{stream}")
 
-                event_count += process_stream(group, stream,
-                                              nil, timestamp, state)
+              # See if we have some stored state for this group and stream.
+              # If we have then use the stored forward_token to pick up
+              # from that point. Otherwise start from the start.
+
+              begin
+                event_count += process_stream(group,
+                                              stream,
+                                              state.store[group][stream]['token'],
+                                              @event_start_time,
+                                              state)
+              rescue Aws::CloudWatchLogs::Errors::InvalidParameterException
+                metric(:increment, 'api.calls.getlogevents.invalid_token')
+                log.error('cloudwatch token is expired or broken. trying with timestamp.')
+
+                # try again with timestamp instead of forward token
+                begin
+                  timestamp = state.store[group][stream]['timestamp']
+                  timestamp = @event_start_time unless timestamp
+
+                  event_count += process_stream(group,
+                                                stream,
+                                                nil,
+                                                timestamp,
+                                                state)
+                rescue => boom
+                  log.error("Unable to retrieve events for stream #{stream} in group #{group}: #{boom.inspect}") # rubocop:disable all
+                  metric(:increment, 'api.calls.getlogevents.failed')
+                  sleep @error_interval
+                  next
+                end
               rescue => boom
-                log.error("Unable to retrieve events for stream #{stream} "\
-                          "in group #{group}: #{boom.inspect}") # rubocop:disable all
+                log.error("Unable to retrieve events for stream #{stream} in group #{group}: #{boom.inspect}") # rubocop:disable LineLength
                 metric(:increment, 'api.calls.getlogevents.failed')
                 sleep @error_interval
                 next
               end
-            rescue => boom
-              log.error("Unable to retrieve events for stream #{stream} in group #{group}: #{boom.inspect}") # rubocop:disable LineLength
-              metric(:increment, 'api.calls.getlogevents.failed')
-              sleep @error_interval
-              next
             end
-          end
+          end while !@log_streams_next_token.nil? && !@finished # don't try to get and process all streams at once
```

Comment:
save the next token for the log_streams globally... if we expose it out like this, we make sure we only process as many log_streams as they have been fetched from the method call